crates/bevy_ecs/src/event/mut_iterators.rs

```diff
@@ -0,0 +1,275 @@
+use crate as bevy_ecs;
+#[cfg(feature = "multi_threaded")]
+use bevy_ecs::batching::BatchingStrategy;
+use bevy_ecs::event::{Event, EventId, EventInstance, Events, ManualEventMutator};
+use bevy_utils::detailed_trace;
+use std::{iter::Chain, slice::IterMut};
+
+/// An iterator that yields any unread events from an [`EventMutator`] or [`ManualEventMutator`].
+#[derive(Debug)]
+pub struct EventMutatorIterator<'a, E: Event> {
+    iter: EventMutatorIteratorWithId<'a, E>,
+}
+
+impl<'a, E: Event> Iterator for EventMutatorIterator<'a, E> {
+    type Item = &'a mut E;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|(event, _)| event)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+
+    fn count(self) -> usize {
+        self.iter.count()
+    }
+
+    fn last(self) -> Option<Self::Item>
+    where
+        Self: Sized,
+    {
+        self.iter.last().map(|(event, _)| event)
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        self.iter.nth(n).map(|(event, _)| event)
+    }
+}
+
+impl<'a, E: Event> ExactSizeIterator for EventMutatorIterator<'a, E> {
+    fn len(&self) -> usize {
+        self.iter.len()
+    }
+}
+
+/// An iterator that yields any unread events (and their IDs) from an [`EventMutator`] or [`ManualEventMutator`].
+#[derive(Debug)]
+pub struct EventMutatorIteratorWithId<'a, E: Event> {
+    mutator: &'a mut ManualEventMutator<E>,
+    chain: Chain<IterMut<'a, EventInstance<E>>, IterMut<'a, EventInstance<E>>>,
+    unread: usize,
+}
+
+impl<'a, E: Event> EventMutatorIteratorWithId<'a, E> {
+    /// Creates a new iterator that yields any `events` that have not yet been seen by `mutator`.
+    pub fn new(mutator: &'a mut ManualEventMutator<E>, events: &'a mut Events<E>) -> Self {
+        let a_index = mutator
+            .last_event_count
+            .saturating_sub(events.events_a.start_event_count);
+        let b_index = mutator
+            .last_event_count
+            .saturating_sub(events.events_b.start_event_count);
+        let a = events.events_a.get_mut(a_index..).unwrap_or_default();
+        let b = events.events_b.get_mut(b_index..).unwrap_or_default();
+
+        let unread_count = a.len() + b.len();
+
```

Comment:
was there a reason to remove the `debug_assert_eq!(unread_count, reader.len(events)`.(from the readonly version)?

Reply:
Yeah, its provably true in all cases that I could think of and length is covered in the existing tests.